diff --git a/arch/x86/boot/compressed/kaslr.c b/arch/x86/boot/compressed/kaslr.c
index a63fbc25ce84..d7dec3310553 100644
--- a/arch/x86/boot/compressed/kaslr.c
+++ b/arch/x86/boot/compressed/kaslr.c
@@ -40,6 +40,9 @@
 #include <linux/ctype.h>
 #include <linux/efi.h>
 #include <generated/utsrelease.h>
+#undef CONFIG_CGROUPS
+#undef CONFIG_MEMCG
+#undef CONFIG_SOCK_CGROUP_DATA
 #include <asm/efi.h>
 
 /* Macros used by the included decompressor code below. */
