diff --git a/net/core/sock.c b/net/core/sock.c
index ec6eb546..e8b0079b 100644
--- a/net/core/sock.c
+++ b/net/core/sock.c
@@ -929,7 +929,7 @@ int sock_setsockopt(struct socket *sock, int level, int optname,
 
 	case SO_ATTACH_BPF:
 		ret = -EINVAL;
-		if (optlen == sizeof(u32)) {
+		if (optlen == sizeof(u32) && likely(sk->sk_family != PF_UNIX)) {
 			u32 ufd;
 
 			ret = -EFAULT;
@@ -955,7 +955,7 @@ int sock_setsockopt(struct socket *sock, int level, int optname,
 
 	case SO_ATTACH_REUSEPORT_EBPF:
 		ret = -EINVAL;
-		if (optlen == sizeof(u32)) {
+		if (optlen == sizeof(u32) && likely(sk->sk_family != PF_UNIX)) {
 			u32 ufd;
 
 			ret = -EFAULT;
