diff --git a/fs/exec.c b/fs/exec.c
index 0da4d748..66cbadd1 100644
--- a/fs/exec.c
+++ b/fs/exec.c
@@ -831,6 +831,8 @@ EXPORT_SYMBOL(transfer_args_to_stack);
 static struct file *do_open_execat(int fd, struct filename *name, int flags)
 {
 	struct file *file;
+	struct inode *inode;
+	uid_t uid;
 	int err;
 	struct open_flags open_exec_flags = {
 		.open_flag = O_LARGEFILE | O_RDONLY | __FMODE_EXEC,
@@ -851,7 +853,8 @@ static struct file *do_open_execat(int fd, struct filename *name, int flags)
 		goto out;
 
 	err = -EACCES;
-	if (!S_ISREG(file_inode(file)->i_mode))
+	inode = file_inode(file);
+	if (!S_ISREG(inode->i_mode))
 		goto exit;
 
 	if (path_noexec(&file->f_path))
@@ -861,6 +864,15 @@ static struct file *do_open_execat(int fd, struct filename *name, int flags)
 	if (err)
 		goto exit;
 
+	if (ns_capable_noaudit(&init_user_ns, CAP_SYS_ADMIN)) {
+		uid = from_kuid(&init_user_ns, inode->i_uid);
+		if (uid || (!uid &&
+			((inode->i_mode & S_IWGRP) || (inode->i_mode & S_IWOTH)))) {
+			err = -EPERM;
+			goto exit;
+		}
+	}
+
 	if (name->name[0] != '\0')
 		fsnotify_open(file);
 
