diff --git a/fs/proc/namespaces.c b/fs/proc/namespaces.c
index 59b17e509f46..c2da8ad6e768 100644
--- a/fs/proc/namespaces.c
+++ b/fs/proc/namespaces.c
@@ -47,6 +47,9 @@ static const char *proc_ns_get_link(struct dentry *dentry,
 	if (!dentry)
 		return ERR_PTR(-ECHILD);
 
+	if (!capable(CAP_SYS_ADMIN))
+		return ERR_PTR(-EPERM);
+
 	task = get_proc_task(inode);
 	if (!task)
 		return error;
@@ -68,6 +71,10 @@ static int proc_ns_readlink(struct dentry *dentry, char __user *buffer, int bufl
 	char name[50];
 	int res = -EACCES;
 
+	if (!capable(CAP_SYS_ADMIN))
+		res = readlink_copy(buffer, buflen, "masked");
+		return res;
+
 	task = get_proc_task(inode);
 	if (!task)
 		return res;
