diff --git a/kernel/ptrace.c b/kernel/ptrace.c
index 84b136793..7786540b9 100644
--- a/kernel/ptrace.c
+++ b/kernel/ptrace.c
@@ -45,6 +45,10 @@ int ptrace_access_vm(struct task_struct *tsk, unsigned long addr,
 	if (!mm)
 		return 0;
 
+	// fcicq: forbid PTRACE_POKETEXT / PTRACE_POKEDATA
+	if (!capable(CAP_SYS_ADMIN) && (gup_flags & FOLL_WRITE))
+		return 0;
+
 	if (!tsk->ptrace ||
 	    (current != tsk->parent) ||
 	    ((get_dumpable(mm) != SUID_DUMP_USER) &&
